### DISCUSSION TITLE
SEP-40 Oracle Rates Consumption Example

### DISCUSSION THREAD

**Discussion by antb123** [stellar/sorobounty-spectacular#18] [General]
Hi Stellarfam,

We are building [LightEcho](https://lightecho.io), which is the Stellar Soroban Oracle for emerging market rates and XLM volatility. It will help developers and support Stellar’s growth as a low-friction, inclusive, global financial network. Now, in light of the [approved merge](https://github.com/stellar/stellar-protocol/pull/1357) of our  [SEP-40](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0040.md)  (A standard interface for oracles transmitting price information) change (thanks!), and to demonstrate our continued devotion to Soroban's growth and progression, we would like to provide an example contract that implements up-and-down price checks against contracts like our own.

### What we made:
We made PriceUpDown, a contract that fetches prices (XLM/USD) from a Soroban [SEP-40](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0040.md) Oracle contract, and returns information if the price went up, down, or remained the same since the last time checked:

```
UpDown {
    up: false,
    down: false,
    equal: false,
}
```

This contract has two main purposes: to demonstrate the capabilities of Soroban price Oracles, and serve as a reference contract code for future Oracle consumer contract developers who want to get familiar with how to interact with [SEP40](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0040.md)-compatible Soroban price Oracles.

### The contract demonstrates:
- Contract to contract communication in Rust
- Storage of a rate using Rust
- Python communicating with a SEP-40 compliant Smart contract (alpha)
- Javascript communication to a SEP-40 contract is in the LightEcho github repo
- In the oracle LightEcho repo we display Soroban authentication

The PriceUpDown contract can be found in our [public GitHub repository](https://github.com/bp-ventures/lightecho-stellar-oracle/tree/trunk/oracle-onchain/sep40/examples/price_up_down).

### Current limitations:
Due to the latest Python Soroban SDK updates in [pre-release v9.0.0-alpha1 ](https://github.com/StellarCN/py-stellar-base/commit/1be4099bb4a73f78e09e31b92546c853cfd465f1)(August 28th, 2023), some features of our CLI (used to invoke the contracts) are raising errors. At the moment the only reliable way of invoking the contract is using Rust ([tests](https://github.com/bp-ventures/lightecho-stellar-oracle/blob/trunk/oracle-onchain/sep40/examples/price_up_down/src/test.rs) for reference).

#### Price Up-Down Contract and Instructions:
[https://github.com/bp-ventures/lightecho-stellar-oracle/tree/trunk/oracle-onchain/sep40/examples/price_up_down](https://github.com/bp-ventures/lightecho-stellar-oracle/tree/trunk/oracle-onchain/sep40/examples/price_up_down)

#### LightEcho Github:
[https://github.com/bp-ventures/lightecho-stellar-oracle/blob/trunk/README.md](https://github.com/bp-ventures/lightecho-stellar-oracle/blob/trunk/README.md)

#### Older JS fiddle
[https://playcode.io/1532409](https://playcode.io/1532409)

### About BPV
[BPV](https://bpventures.us) is a leading fintech and blockchain consulting firm, specializing in empowering financial institutions to use digital assets to optimize their operations and achieve revenue growth. BPV has extensive expertise with the Stellar protocol, regulatory compliance, and IT systems, and is building the future of decentralized finance by developing software solutions for secure digital asset custody and management and tokenized asset trading including smart contracts. For media inquiries, please contact: support@bpventures.us

**Comment by briwylde08** [comment]
Amazing! Thank you so much for your submission! We’ll review it and be in contact with you within the next week. 🎉

**Comment by Julian-dev28** [comment]
Hey @antb123 thanks for the submission! We really enjoy the work the lightecho team has produced throughout SCF!

Re setup:
I've successfully deployed the contract however when I attempt to initialize:

```
$REPO_DIR/oracle-onchain/sep40/cli/cli oracle initialize \
  GDOOLD2UL3STZ4FLHM5CV3ZFSTYI4EYZHEEGIC4GHL4CJ4BLSSYNN5ER \
  XLM \
  18 \
  300
```

 I receive the following error:

```
Traceback (most recent call last):
  File "/Users/julianmartinez/Documents/GitHub/lightecho-stellar-oracle/oracle-onchain/sep40/cli/cli.py", line 10, in <module>
    from colorama import init as colorama_init
ModuleNotFoundError: No module named 'colorama'
```

I've tried using pip etc but nothing has worked so far. I am also not the best with python so it may be a user issue.
Are you able to add directions to the readme to install dependencies? 

Also RE:

> Due to the latest Python Soroban SDK updates in [pre-release v9.0.0-alpha1 ](https://github.com/StellarCN/py-stellar-base/commit/1be4099bb4a73f78e09e31b92546c853cfd465f1)(August 28th, 2023), some features of our CLI (used to invoke the contracts) are raising errors. At the moment the only reliable way of invoking the contract is using Rust ([tests](https://github.com/bp-ventures/lightecho-stellar-oracle/blob/trunk/oracle-onchain/sep40/examples/price_up_down/src/test.rs) for reference).

Can you please tell me what I should be on the lookout for? Which invocations work at the moment?

**Comment by antb123** [comment]
Hi we updated the readme (you need to call poetry) as well we updated
information regarding invocations .

Thanks for testing this out!

**Comment by Julian-dev28** [comment]
@antb123 Thanks for the update!

It's still unclear on what steps to take to interact with the contract. Are you able to give more instructions in the [up down contract readme?
](https://github.com/bp-ventures/lightecho-stellar-oracle/blob/trunk/README.md)

Following along step by step, I am running into the error below:

**re: Initialize the Lightecho Oracle**

```
# initialize the oracle
$REPO_DIR/oracle-onchain/sep40/cli/cli oracle initialize \
  GDOOLD2UL3STZ4FLHM5CV3ZFSTYI4EYZHEEGIC4GHL4CJ4BLSSYNN5ER \
  XLM \
  18 \
  300

Traceback (most recent call last):
  File "/Users/julianmartinez/Documents/GitHub/g/lightecho-stellar-oracle/oracle-onchain/sep40/cli/cli.py", line 10, in <module>
    from colorama import init as colorama_init
ModuleNotFoundError: No module named 'colorama'
  ```
  
It would be great to tell users if they are supposed to install python dependencies here and how to do it _or_ direct them to another Readme from the repo that specifies this.

**Comment by yuriescl** [comment]
@Julian-dev28 Hello, we apologize for those small issues.

The documentation has been updated with the instructions that were missing before:
https://github.com/bp-ventures/lightecho-stellar-oracle/blob/trunk/oracle-onchain/sep40/examples/price_up_down/README.md

There is an issue ([GitHub issue](https://github.com/StellarCN/py-stellar-base/issues/773)) when invoking the `add_price` function, we're trying to fix it ASAP.

**Reply by Julian-dev28** [reply]
Perfect! I was able to execute some of the functions from the and get the return data so that is great. 

Feel free to ping me when you get that fix in and I'll be happy to do a final review!

**Comment by yuriescl** [comment]
@Julian-dev28 The issue has been identified and fixed. It was incorrect source account in the Python CLI code. Please `git pull` and retry, let us know if there's any further issue

**Reply by Julian-dev28** [reply]
Hey @yuriescl 

I was able to follow the tutorial successfully all the way through until the very last step:

```
$ ./cli priceupdown get_price_up_down \
  other \
  USD
Output:
Traceback (most recent call last):

  File "/Users/julianmartinez/Documents/GitHub/g/lightecho-stellar-oracle/oracle-onchain/sep40/cli/cli.py", line 628, in <module>
    app()

  File "/Users/julianmartinez/Documents/GitHub/g/lightecho-stellar-oracle/oracle-onchain/sep40/cli/cli.py", line 602, in priceupdown_get_price_up_down
    invoke_and_output(

  File "/Users/julianmartinez/Documents/GitHub/g/lightecho-stellar-oracle/oracle-onchain/sep40/cli/cli.py", line 247, in invoke_and_output
    output_tx_data(tx_data)

  File "/Users/julianmartinez/Documents/GitHub/g/lightecho-stellar-oracle/oracle-onchain/sep40/cli/cli.py", line 215, in output_tx_data
    print(parse_sc_map(result.map.sc_map))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/Users/julianmartinez/Documents/GitHub/g/lightecho-stellar-oracle/oracle-onchain/sep40/cli/cli.py", line 200, in parse_sc_map
    value = parse_sc_val(entry.val)
            ^^^^^^^^^^^^^^^^^^^^^^^

  File "/Users/julianmartinez/Documents/GitHub/g/lightecho-stellar-oracle/oracle-onchain/sep40/cli/cli.py", line 186, in parse_sc_val
    raise ValueError("Could not parse sc_val")

ValueError: Could not parse sc_val
```

Once this is fixed, I can do the final review and pass this along to Bri!

**Reply by yuriescl** [reply]
@Julian-dev28  Hello, the issue has been fixed, please `git pull` and retry:
```
$ ./cli priceupdown get_price_up_down other USD
Output:
{'down': False, 'equal': True, 'up': False}
Horizon tx:
https://horizon-futurenet.stellar.org/transactions/585b700031848ccc650c449e1898f27344becca979697f
6c4b53c2586f2cbcbb
```

**Reply by Julian-dev28** [reply]
Thanks for the update!

Everything looks good so I believe I can wrap the review process and hand this over to Bri!

Awesome use case here and I am happy to have reviewed it!

Just be sure to upgrade with the latest preview release to stay up to date with our latest functionality 🚀

**Reply by yuriescl** [reply]
Thanks! Yeah there were some breaking changes in latest Soroban update, we're working on fixing the issues.

**Comment by briwylde08** [comment]
Hi @antb123! Thank you so much again for your submission and discussion! [🎉](https://emojipedia.org/party-popper/)

We had a chance to review it completely earlier and everything looked great. However, looks like you might need to upgrade again because of the recent Futurenet upgrade.

Given our guidelines, your submission may qualify for an award of $450 worth of XLM. To receive that award, please fill out [this form](https://stellarform.typeform.com/to/HBjEyCJt) completely and accurately.

Thanks again! Please let us know if you have any questions.

Best,
Bri

**Comment by antb123** [comment]
I filled it in as BP Ventures and submitted the required tax information.

I wasn't sure if you needed us to resubmit full KYC as we did it 3 months
ago. If you do, I will resubmit the form tomorrow.

Thank you

**Reply by briwylde08** [reply]
Hi there! No need to fill out the form again!

We've logged your award payment of $450 worth of XLM, which will be sent to the wallet you designated in the next week.

Thanks again and please let me know if you have any questions!

Best,
Bri
